src/otx/algorithms/visual_prompting/tasks/openvino.py

```diff
@@ -338,84 +414,125 @@ def predict(self, dataset_item: DatasetItemEntity) -> List[Annotation]:  # type:
+                predicted_masks[label].append(prediction[self.model["decoder"].output_blob_name])
                 used_points[label].append(points_score)
-        self.__inspect_overlapping_areas(predicted_masks, used_points, annotations)
-        return sum(annotations.values(), [])
 
-    def forward_prompt_getter(self, inputs: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+        self._inspect_overlapping_areas(predicted_masks, used_points)
+
+        for label, predictions in predicted_masks.items():
+            if len(predictions) == 0:
+                continue
+            metadata = {
+                "label": [_label for _label in self.labels if int(_label.id_) == label][0],
+                "original_size": original_size,
+            }
+            for prediction, used_point in zip(predictions, used_points[label]):
+                annotation, _, _ = self.post_process(
+                    {self.model["decoder"].output_blob_name: prediction, "scores": used_point[-1]}, metadata
+                )
+                annotations[label].extend(annotation)
+
+        return sum(annotations.values(), []), predicted_masks, used_points
+
+    def forward_prompt_getter(
+        self,
+        image_embeddings: Dict[str, np.ndarray],
+        reference_feats: np.ndarray,
+        used_indices: np.ndarray,
+        original_size: np.ndarray,
+    ) -> Tuple[Dict[int, np.ndarray], Dict[int, np.ndarray]]:
         """Forward function of OpenVINO Visual Prompting Inferencer."""
-        return self.model["prompt_getter"].infer_sync(inputs)
+        inputs = {
+            "original_size": original_size,
+            "threshold": np.array([[self.model["prompt_getter"].sim_threshold]], dtype=np.float32),
+            "num_bg_points": np.array([[self.model["prompt_getter"].num_bg_points]], dtype=np.int64),
+            **image_embeddings,
+        }
+        total_points_scores: Dict[int, np.ndarray] = {}
+        total_bg_coords: Dict[int, np.ndarray] = {}
+        for label in used_indices[0]:
+            reference_feat = reference_feats[label]
+            inputs["reference_feat"] = reference_feat
+            outputs = self.model["prompt_getter"].infer_sync(inputs)
+
+            total_points_scores[label] = outputs["points_scores"]
+            total_bg_coords[label] = outputs["bg_coords"]
+
+        return total_points_scores, total_bg_coords
 
     def forward_decoder(  # type: ignore
-        self, inputs: Dict[str, np.ndarray], original_size: np.ndarray
+        self,
+        inputs: Dict[str, np.ndarray],
+        original_size: np.ndarray,
+        is_cascade: bool = True,
     ) -> Dict[str, np.ndarray]:
         """Forward function of OpenVINO Visual Prompting Inferencer."""
+        masks: np.ndarray
         logits: np.ndarray
         scores: np.ndarray
-        mask_slice = slice(0, 1)
-        for i in range(3):
+        num_iter = 3 if is_cascade else 1
+        for i in range(num_iter):
```

Comment:
how about dropping for loop here?
do first-step prediction
if is_cascade:
    perform 2 times of post-refinement

Reply:
Then, we must repeatedly implement this part.
https://github.com/openvinotoolkit/training_extensions/blob/b3448f67471fcff6e21b52de6d5b22756eef89f7/src/otx/algorithms/visual_prompting/tasks/openvino.py#L520-L527